File: source/GeneBase.ts
/*!
 * @author electricessence / https://github.com/electricessence/
 * Licensing: MIT https://github.com/electricessence/Genetic-Algorithm-Platform/blob/master/LICENSE.md
 */


import {Lazy} from "typescript-dotnet-umd/System/Lazy";
import {List} from "typescript-dotnet-umd/System/Collections/List";
import {ArgumentException} from "typescript-dotnet-umd/System/Exceptions/ArgumentException";
import {IEquatable} from "typescript-dotnet-umd/System/IEquatable";
import {IGene} from "./IGene";
import {ILinqEnumerable} from "typescript-dotnet-umd/System.Linq/Enumerable";

abstract class GeneBase<T extends IGene>
extends List<T> implements IGene, IEquatable<GeneBase<T>>
{
	constructor()
	{
		super();
		this.resetToString();
	}

	abstract serialize():string;

	abstract clone():GeneBase<T>;

	get descendants():ILinqEnumerable<IGene>
	{
		const e:ILinqEnumerable<IGene> = this.linq;
		return e.concat(e.selectMany(s=>s.descendants));
	}

	findParent(child:T):IGene|null
	{
		let children = this._source;
		if(!children || !children.length) return null;
		if(children.indexOf(child)!= -1) return this;

		for(let c of children)
		{
			let p = c.findParent(child);
			if(p) return p;
		}

		return null;
	}

	protected _replaceInternal(target:T, replacement:T, throwIfNotFound?:boolean):boolean
	{
		const s = this._source;
		const index = this._source.indexOf(target);
		if(index== -1)
		{
			if(throwIfNotFound)
				throw new ArgumentException('target', "gene not found.");
			return false;
		}

		s[index] = replacement;
		return true;
	}

	replace(target:T, replacement:T, throwIfNotFound?:boolean):boolean
	{
		const m = this._replaceInternal(target, replacement, throwIfNotFound);
		if(m) this._onModified();
		return m;
	}

	_toString:Lazy<string>;

	resetToString():void
	{
		const ts = this._toString;
		if(ts) ts.tryReset();
		else this._toString = new Lazy<string>(()=>this.toStringInternal(), false, true);
		
		this.forEach(c=>c.resetToString());
	}

	protected _onModified()
	{
		super._onModified();
		this.resetToString();
	}

	protected abstract toStringInternal():string;

	toString()
	{
		return this._toString.value;
	}

	equals(other:GeneBase<T>):boolean
	{
		return this===other || this.toString()==other.toString();
	}
}

export default GeneBase;

File: source/GeneBase.cs
/*!
 * @author electricessence / https://github.com/electricessence/
 * Licensing: MIT https://github.com/electricessence/Genetic-Algorithm-Platform/blob/master/LICENSE.md
 */

using System;
using System.Linq;
using System.Collections.Generic;
using System.Threading;
using System.Collections;
using Open.Threading;
using Open.Collections;
using Open;

namespace GeneticAlgorithmPlatform
{

	public abstract class GeneBase : ModificationSynchronizedBase, IGene
	{

		Lazy<string> _toString;
		public virtual void ResetToString()
		{
			if (_toString == null || _toString.IsValueCreated)
				_toString = Lazy.New(ToStringInternal);
		}

		protected override void OnDispose(bool calledExplicitly)
		{
			base.OnDispose(calledExplicitly);
			_toString = null;
		}

		protected override void OnModified()
		{
			base.OnModified();
			ResetToString();
		}

		protected abstract string ToStringInternal();

		public sealed override string ToString()
		{
			return _toString == null ? base.ToString() : _toString.Value;
		}

		public virtual string Serialize()
		{
			return ToString();
		}

		public virtual IGene Clone()
		{
			return Sync.Reading(() => CloneInternal());
		}

		protected abstract IGene CloneInternal();

		public virtual bool Equivalent(IGene other)
		{
			return this == other || other != null && this.ToString() == other.ToString();
		}

		protected override void OnFrozen()
		{
			base.OnFrozen();
			ResetToString();
		}

	}

	public abstract class GeneBase<T> : GeneBase, IGeneNode<T>
	where T : IGene
	{

		protected override void OnDispose(bool calledExplicitly)
		{
			base.OnDispose(calledExplicitly);
			var c = _children as IDisposable;
			_children = null;
			c.SmartDispose();
			_getChildNodes = null;
			_getDescendants = null;
		}

		override protected ModificationSynchronizer InitSync(object sync = null)
		{
			ModificationSynchronizer s;
			_children = new ThreadSafeTrackedList<T>(out s);
			return s;
		}

		private ThreadSafeTrackedList<T> _children;

		public IEnumerable<T> Children
		{
			get
			{
				return GetChildren();
			}
		}

		protected ThreadSafeTrackedList<T> GetChildren()
		{
			// Since we can't know if changes occur in our children, we have to return the uncached value until it is set to read-only.
			var c = _children;
			AssertIsLiving();
			return c;
		}

		IEnumerable<IGeneNode<T>> _getChildNodes;
		public IEnumerable<IGeneNode<T>> ChildNodes
		{
			get
			{
				var children = GetChildren();
				return LazyInitializer.EnsureInitialized(ref _getChildNodes,
					() => children.Where(c => c is IGeneNode<T>).Cast<IGeneNode<T>>());
			}
		}

		public int Count
		{
			get
			{
				return GetChildren().Count;
			}
		}


		IEnumerable<T> _getDescendants;
		protected IEnumerable<T> GetDescendants()
		{
			var children = GetChildren();
			return LazyInitializer.EnsureInitialized(ref _getDescendants,
				() => children.Concat(ChildNodes.SelectMany(s => s.Descendants)));
		}

		Lazy<T[]> _descendants;
		public IEnumerable<T> Descendants
		{
			get
			{
				var d = _descendants;
				AssertIsLiving();
				// Since we can't know if changes occur in our children, we have to return the uncached value until it is set to read-only.
				return d==null ? GetDescendants() : d.Value;
			}
		}

		public virtual void Add(T item)
		{
			GetChildren().Add(item);
		}

		public void Add(T item, T item2, params T[] items)
		{
			GetChildren().AddThese(new T[] { item, item2 }.Concat(items));
		}

		public void AddThese(IEnumerable<T> items)
		{
			GetChildren().AddThese(items);
		}

		public void Clear()
		{
			GetChildren().Clear();
		}

		public bool Contains(T item)
		{
			return GetChildren().Contains(item);
		}

		public void CopyTo(T[] array, int arrayIndex)
		{
			GetChildren().CopyTo(array, arrayIndex);
		}

		public IGeneNode<T> FindParent(T child)
		{
			return Sync.Reading(() =>
			{
				var children = GetChildren();
				if (children.Count == 0) return null;
				if (children.IndexOf(child) != -1) return this;

				foreach (var c in ChildNodes)
				{
					var p = c.FindParent(child);
					if (p != null) return p;
				}

				return null;
			});

		}

		public bool Remove(T item)
		{
			return GetChildren().Remove(item);
		}


		public IEnumerator<T> GetEnumerator()
		{
			return GetChildren().GetEnumerator();
		}

		public bool ReplaceChild(T item, T replacement, bool throwIfNotFound = false)
		{
			return GetChildren().Replace(item, replacement, throwIfNotFound);
		}

		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetChildren().GetEnumerator();
		}

		public new IGeneNode<T> Clone()
		{
			return Sync.Reading(() => (IGeneNode<T>)CloneInternal());
		}

		protected override void OnFrozen()
		{
			base.OnFrozen();
			var children = GetChildren();
			children.Freeze();
			foreach (var child in children)
				child.Freeze();

			_descendants = Lazy.New(() => GetDescendants().ToArray());
		}


	}

}